### PR TITLE
feat(stream): add ordered CDC snapshot diff reader

### PR DESCRIPTION
**Pull request by hzxa21**
Tracking issue: #26481
Stacked on: #26444

Add the dormant compute-side reader used to compare a fresh upstream CDC snapshot with a committed RisingWave table snapshot.

- Merge the upstream and RisingWave snapshots in primary-key order.
- Emit insert, delete, and upsert corrections while skipping equal rows.
- Emit diff-local progress and completion events for later recovery wiring.
- Support reading the existing table from a pinned committed epoch.
- Preserve full primary-key ordering for both prefix and non-prefix split keys.

This PR only introduces the reusable reader. It does not modify either existing CDC backfill executor, expose `ALTER TABLE ... RESNAPSHOT`, or start RESNAPSHOT orchestration.

### CODE REVIEW

**Comment by hzxa21** [comment]
This change is part of the following stack:

- #26444
    - #26482 ◀
        - #26521

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+}
+
+impl<S: StateStore> ResnapshotDiffRead<S> {
+    pub fn new(
+        state_store: S,
+        table_desc: &StorageTableDesc,
+        output_column_ids: &[ColumnId],
+        read_epoch: u64,
+    ) -> Self {
+        let vnodes = Some(Bitmap::ones(table_desc.vnode_count()).into());
```
Comment (hzxa21) [inline src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs:51]: - cdc backfill is a singleton so we should use all vnodes initialize the batch table
- cdc backfill v2 is not singletone but it is range partition in terms of the upstream table read range so the batch table also needs to consider all vnodes.

File: src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+        diff_ordered_row_streams(
+            left,
+            right,
+            pk_indices,
+            pk_order,
+            compare_indices,
+            data_types,
+            chunk_size,
+        )
+    }
```
Comment (hzxa21) [inline src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs]: We are not going to support parallel backfill (i.e. cdc backfill v2) at the beginning. We can avoid introducing `snapshot_read_table_split_diff` and `storage_table_split_read` in this PR.

File: src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+                    &read_args.left_bound_inclusive,
+                    &read_args.right_bound_exclusive,
+                )?
+            {
+                continue;
+            }
+            yield row;
+        }
+    }
+}
```
Comment (hzxa21) [inline src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs]: We are not going to support parallel backfill (i.e. cdc backfill v2) at the beginning. We can avoid introducing `snapshot_read_table_split_diff` and `storage_table_split_read` in this PR.

File: src/stream/src/executor/backfill/cdc/upstream_table/diff.rs [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+    indices
+        .iter()
+        .all(|idx| left.datum_at(*idx) == right.datum_at(*idx))
+}
+
+fn project_pk(row: &OwnedRow, pk_indices: &[usize]) -> OwnedRow {
+    row.project(pk_indices).into_owned_row()
+}
+
+struct DiffChunkBuilder<'a> {
```
Comment (hzxa21) [inline src/stream/src/executor/backfill/cdc/upstream_table/diff.rs]: Any specific reason on why we don't use the stream chunk builder directlu but use the diff chunk builder wrapper?

File: src/stream/src/executor/backfill/cdc/upstream_table/diff.rs [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+                        output.push(Op::Insert, left);
+                        for output in output.take_flushed_chunks() {
+                            yield output;
+                        }
+                        pending_progress = None;
+                        skipped_since_progress = 0;
+                    } else {
+                        pending_progress = Some(project_pk(&left, &pk_indices));
+                        skipped_since_progress += 1;
+                        if skipped_since_progress >= chunk_size {
```
Comment (hzxa21) [inline src/stream/src/executor/backfill/cdc/upstream_table/diff.rs]: It can be a potential starvation here if we only yield when skipped_since_progress >= chunk_size.
1. diff_ordered_row_streams is interrupted by whatever reason in the caller side when skipped_since_progress < chunk_size
2. On the next round, diff_ordered_row_streams will be reset to the previous pk position
3. it can interrupted again before the pk progress has been made, repeat to 1.

File: src/stream/src/executor/backfill/cdc/upstream_table/diff.rs [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+
+fn compare_pk(
+    left: &OwnedRow,
+    right: &OwnedRow,
+    pk_indices: &[usize],
+    pk_order: &[OrderType],
+) -> Ordering {
+    let left_pk = left.project(pk_indices);
+    let right_pk = right.project(pk_indices);
+    cmp_datum_iter(left_pk.iter(), right_pk.iter(), pk_order.iter().copied())
```
Comment (hzxa21) [inline src/stream/src/executor/backfill/cdc/upstream_table/diff.rs]: **[High] Please make this comparison connector-aware (or reject unsupported PK orderings).** MySQL `BIGINT UNSIGNED` values above `i64::MAX` are represented as negative `i64` values in RisingWave. The upstream stream is therefore ordered as `[5, -1]`, while the storage scan is ordered as `[-1, 5]`. Merging them with `cmp_datum_iter` can produce `Delete(-1) -> Progress(5) -> Insert(-1)`, allowing a buffered CDC event for `-1` to be released before the stale snapshot insert and then overwritten. The existing CDC backfill obtains `pk_column_unsigned_i64_compare_flags` and uses `cmp_pk_unsigned_aware`; this reader needs the same semantics. Please also add a regression test across the signed/unsigned boundary.

**Review by hzxa21 — COMMENTED** [review]
Three correctness/operability findings from the ordered resnapshot diff review.

File: src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        read_args: SnapshotReadArgs,
+        batch_size: u32,
+        pk_order: Vec<OrderType>,
+        compare_indices: Vec<usize>,
+        data_types: Vec<DataType>,
+        chunk_size: usize,
+    ) -> impl Stream<Item = StreamExecutorResult<SnapshotReadOutput>> + Send + 'a {
+        let pk_indices = read_args.pk_indices.clone();
+        let left = snapshot_read_rows(
+            upstream_table_reader.snapshot_read_full_table(read_args.clone(), batch_size),
```
Comment (hzxa21) [inline src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs]: **[High] This diff path needs strict external-row decoding.** The existing PostgreSQL, MySQL, and SQL Server snapshot converters log conversion failures and return SQL `NULL`. That was already risky for initial backfill, but treating the decoded row as authoritative input to a destructive resnapshot makes the consequence worse: a non-PK failure overwrites a correct stored value with `NULL`, while a PK failure breaks the merge ordering and can classify correct storage rows as right-only deletes. Please propagate conversion errors in this path, reject null decoded PKs, and add tests for malformed PK and non-PK values.

File: src/stream/src/executor/backfill/cdc/upstream_table/diff.rs [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+                output.push(Op::Insert, left_row.take().expect("checked left row"));
+                for output in output.take_flushed_chunks() {
+                    yield output;
+                }
+                pending_progress = None;
+                skipped_since_progress = 0;
+                left_row = left_stream.next().await.transpose()?;
+            }
+            (None, Some(_)) => {
+                output.push(Op::Delete, right_row.take().expect("checked right row"));
```
Comment (hzxa21) [inline src/stream/src/executor/backfill/cdc/upstream_table/diff.rs]: **[Medium] Storage-only deletes bypass the configured snapshot rate limit.** The limiter is applied only while reading rows from the upstream/left stream. If the upstream table is empty, or there is a long stale range before its next key, this branch scans storage and emits delete corrections at full speed. That can create an unbounded burst of compute, state-store writes, and downstream traffic during repair. Please charge right-only deletes to the snapshot limiter (without double-charging inserts/updates) and add an empty-left/many-right regression test.

File: src/stream/src/executor/backfill/cdc/upstream_table/diff.rs (hunk 6 above)
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/diff.rs]: Agreed. Reusing the unsigned-aware comparator alone would still be unsafe because the RisingWave storage scan remains in signed `i64` order. I took the explicitly allowed reject path in 1aedc1d613: the reader obtains the connector flags and fails before merging any `BIGINT UNSIGNED` PK, with a regression covering upstream `[5, -1]` versus storage `[-1, 5]`.

File: src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs (hunk 7 above)
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs]: Agreed. 528d85ee04 adds strict snapshot-row decoders for PostgreSQL, MySQL, and SQL Server while preserving lenient decoding for existing initial backfill. Resnapshot now uses only the strict entry points, propagates non-PK conversion failures, and rejects decoded NULL PKs; malformed PK/non-PK and NULL-PK regressions are included.

File: src/stream/src/executor/backfill/cdc/upstream_table/diff.rs (hunk 8 above)
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/diff.rs]: Agreed. e29a4ec877 shares one snapshot `RateLimiter` between upstream reads and right-only deletes. Deletes consume one permit, while insert/update rows keep the permit already charged by the upstream reader, so they are not double-charged. The empty-left/many-right paused-limiter regression and a no-double-charge regression both pass.

File: src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs (hunk 1 above)
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs:51]: Agreed. The singleton reader already initializes the partial `BatchTable` with `Some(Bitmap::ones(table_desc.vnode_count()).into())`, so the committed storage scan covers every vnode. No code change was needed for this thread; the parallel/v2 path mentioned in the second bullet is now deferred from this initial PR in f107435469.

File: src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs (hunk 2 above)
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs]: Agreed. f107435469 removes `snapshot_read_table_split_diff` and the resnapshot-only strict split plumbing. This initial PR now exposes only the singleton full-table diff reader; parallel CDC backfill support is deferred.

File: src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs (hunk 3 above)
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs]: Agreed. f107435469 removes `storage_table_split_read` together with its split-bound/filter helpers and test. The first version now reads the committed table only through the full-table singleton path.

File: src/stream/src/executor/backfill/cdc/upstream_table/diff.rs (hunk 4 above)
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/diff.rs]: There was no behavior that required the wrapper. 3de3118661 removes `DiffChunkBuilder` and uses `StreamChunkBuilder` directly: `append_row` handles full chunks, while `take` flushes buffered corrections before progress and at end-of-stream.

File: src/stream/src/executor/backfill/cdc/upstream_table/diff.rs (hunk 5 above)
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/diff.rs]: Agreed. 6f45879a39 now emits `Progress` after every equal row, flushing any earlier buffered correction first. This guarantees that cancellation/rebuild can resume from a strictly later PK even when fewer than `chunk_size` rows were skipped. The new `test_streaming_diff_progresses_before_chunk_fills` covers one equal row with `chunk_size = 128`.

File: src/stream/src/executor/backfill/cdc/upstream_table/diff.rs (hunk 4 above)
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/diff.rs]: Follow-up: f94e35d2ef simplifies this further by removing chunk building from the diff reader entirely. The reader now emits row-level `Row { op, row }`, `Progress`, and `Finished` events; the consuming executor owns `StreamChunkBuilder` and its flush policy. The output contract requires all earlier correction rows to be emitted/durable before checkpointing a later progress position.

File: src/stream/src/executor/backfill/cdc/upstream_table/diff.rs (hunk 5 above)
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/diff.rs]: Follow-up after discussing chunk granularity: f94e35d2ef keeps per-equal-row progress but no longer flushes or creates a chunk in the diff reader. It emits row-level correction/progress events. The upper executor can coalesce consecutive progress positions and decide when to build/flush chunks, while preserving the documented rule that progress cannot be checkpointed before earlier correction rows are emitted.

File: src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs (hunk 7 above)
Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs]: Follow-up after separating PK and non-PK semantics: I suggest scoping the new strict behavior to primary-key columns. A real SQL `NULL` is already represented as `Ok(None)`; the compatibility question here is specifically what to do with a conversion error. Initial backfill has historically logged a non-PK conversion error and substituted `NULL`, so resnapshot can preserve that behavior for non-PK columns while decoding every PK strictly and rejecting any null decoded PK before the merge. This removes the ordering/false-delete risk without changing the established non-PK snapshot semantics. The corresponding initial-backfill PK gap is tracked in #26509.

Comment (hzxa21) [reply src/stream/src/executor/backfill/cdc/upstream_table/resnapshot.rs]: Agreed. #26513 now provides the canonical snapshot contract: PK columns decode strictly and decoded-NULL PKs are rejected, while non-PK conversion errors keep the historical log-and-NULL compatibility behavior. Commit 20b9fa81f0 removes #26482’s duplicate full-row strict converters/APIs and its second PK-null validation pass; resnapshot now reuses the ordinary snapshot reader and retains only the shared rate limiter. Verified with 6 strict-PK connector tests, 14 upstream-table tests (1 ignored), connector/stream Clippy with -D warnings, fmt, and diff checks.
